typescript/schedules.md

```diff
@@ -0,0 +1,1095 @@
+ * {step} -> all values, by step (implies start and end are full range for this field)
+ * {end} and {end, step} are not allowed
+ * TODO is there a way to express not allowed ^ in type?
+ * ```
+ *
+ * For example:
+ *
+ * ```
+ * {start: 2, end: 10, step: 3} -> 2, 5, 8
+ * ```
+ */
+export interface Range<Unit> {
+  start?: Unit;
+  end?: Unit;
+
+  /**
+   * The step to take between each value.
+   *
+   * @default 1
+   */
+  step?: PositiveInteger;
+}
+
+export type NumberRange = Range<number> | number
+export type NumberSpec = NumberRange | NumberRange[] | string;
+export type NumberSpecDescription = Range<number>[];
+
+export type Month = 'JANUARY' | 'FEBRUARY' | 'MARCH' | 'APRIL' | 'MAY' | 'JUNE' | 'JULY' | 'AUGUST' | 'SEPTEMBER' | 'OCTOBER' | 'NOVEMBER' | 'DECEMBER';
+export type MonthRange = Range<Month> | Month;
+export type MonthSpec = MonthRange | MonthRange[] | string;
+export type MonthSpecDescription = Range<Month>[];
+
+export type Day = 'SUNDAY' | 'MONDAY' | 'TUESDAY' | 'WEDNESDAY' | 'THURSDAY' | 'FRIDAY' | 'SATURDAY';
+export type DayRange = Range<Day> | Day;
+export type DaySpec = DayRange | DayRange[] | string;
+export type DaySpecDescription = Range<Day>[];
+
+/**
+ * An event specification relative to the calendar, similar to a traditional cron specification.
+ *
+ * A second in time matches if all fields match. This includes `dayOfMonth` and `dayOfWeek`.
+ */
+export interface CalendarSpec {
+  /**
+   * Valid values: 0–59
+   * 
+   * @default 0
+   */
+  second?: NumberSpec;
+
+  /**
+   * Valid values: 0–59
+   * 
+   * @default 0
+   */
+  minute?: NumberSpec;
+
+  /**
+   * Valid values: 0–59
+   * 
+   * @default 0
+   */
+  hour?: NumberSpec;
+
+  /**
+   * Valid values: 1–31
+   * 
+   * @default '*'
+   */
+  dayOfMonth?: NumberSpec;
+
+  /**
+   * @default '*'
+   */
+  month?: MonthSpec;
+
+  /**
+   * Use full years, like `2030`
+   * 
+   * @default '*'
+   */
+  year?: NumberSpec;
+
+  /**
+   * @default '*'
+   */
+  dayOfWeek?: DaySpec;
+}
+
+/**
+ * The version of {@link CalendarSpec} that you get back from {@link ScheduleHandle.describe}
+ */
+export interface CalendarSpecDescription {
+  /**
+   * Valid values: 0–59
+   * 
+   * @default `[{ start: 0 }]`
+   */
+  second?: NumberSpecDescription;
+
+  /**
+   * Valid values: 0–59
+   * 
+   * @default `[{ start: 0 }]`
+   */
+  minute?: NumberSpecDescription;
+
+  /**
+   * Valid values: 0–59
+   * 
+   * @default `[{ start: 0 }]`
+   */
+  hour?: NumberSpecDescription;
+
+  /**
+   * Valid values: 1–31
+   * 
+   * @default `[{ start: 1, end: 31 }]`
+   */
+  dayOfMonth?: NumberSpecDescription;
+
+  /**
+   * @default `[{ start: 'JANUARY' , end: 'DECEMBER' }]`
+   */
+  month?: MonthSpecDescription;
+
+  /**
+   * Use full years, like `2030`
+   * 
+   * @default All possible values
+   */
+  year?: NumberSpecDescription;
+
+  /**
+   * @default `[{ start: 'SUNDAY' , end: 'SATURDAY' }]`
+   */
+  dayOfWeek?: DaySpecDescription;
+}
```

Comment:
@dnr my understanding from Roey is that `DescribeScheduleResponse` will contain Ranges? If yes, what will the default `year` and `dayOfWeek` be?

Reply:
You saw https://github.com/temporalio/api/pull/230, just clarifying here:

The intention is users can supply a calendar-type spec in one of three ways: single cron string, labeled cron string, structured proto. The first two will be validated and converted to structured proto on the frontend, so when you get a schedule back, you'll only see structured proto.

If we add ical rrule support, that may also get compiled into the structured proto, or it might need to be a third major type (calendar, interval, rrule) if it's not expressible that way.

Reply:
What's a labeled cron string?

Reply:
Sorry, that's what I think of CalendarSpec as in my head